scipy/optimize/_qap.py

```diff
@@ -0,0 +1,116 @@
+    score = float(np.real(np.sum(A * B[col_ind][:, col_ind])))
+    return OptimizeResult({"col_ind": col_ind, "score": score})
+
+
+def quadratic_assignment(A, B, maximize=False):
+    """Find an approximate solution to the quadratic assignment problem.
+
+    This function finds approximate solutions to the quadratic assignment
+    problem (QAP) [1]_, which is a problem of the form:
+
+    .. math::
+
+        \\min_P \\& \\ {\\text{trace}(APB^T P^T)}\\\\
+        \\mbox{s.t.} \\& {P \\ \\epsilon \\ \\mathcal{P}}\\\\
+
+    Here :math:`A` and :math:`B` are non-negative square matrices and
+    :math:`\\mathcal{P}` is the set of all permutation matrices.
+
+    The QAP is a NP-hard problem. This function uses Umeyama's
+    eigendecomposition approach [2]_ to find an approximate solution. The
+    results found are not guaranteed to be optimal.
+
+    Parameters
+    ----------
+    A : 2d-array, square, non-negative
+        A square matrix, also called the *cost* matrix.
+
+    B : 2d-array, square, non-negative
+        A square adjacency matrix, also called the *distance* matrix.
+
+    maximize : bool (default: False)
+        Maximizes the objective function if true.
+
+    Returns
+    -------
+    res : OptimizeResult
+        A :class:`scipy.optimize.OptimizeResult` consisting of the fields:
+            col_ind : array
+                An array of column indices giving the assignment.
+            score : float
+                The objective value of the assignment.
+
+    References
+    ----------
+
+    .. [1] https://en.wikipedia.org/wiki/Quadratic_assignment_problem
+
+    .. [2] S Umeyama. An eigendecomposition approach to weighted graph matching
+           problems. *IEEE Transactions on Pattern Analysis and Machine
+           Intelligence* 10(5):695 - 703, September 1988,
+           https://doi.org/10.1109/34.6778
+
+    Examples
+    --------
+
+    >>> A = np.array([[0, 5, 8, 6],
+    ...               [5, 0, 5, 1],
+    ...               [8, 5, 0, 2],
+    ...               [6, 1, 2, 0]])
+    >>> B = np.array([[0, 1, 8, 4],
+    ...               [1, 0, 5, 2],
+    ...               [8, 5, 0, 5],
+    ...               [4, 2, 5, 0]])
+    >>> from scipy.optimize import quadratic_assignment
+    >>> res = quadratic_assignment(A, B)
+    >>> print(res)
+     col_ind: array([3, 2, 1, 0])
+       score: 200.0
+
+    The `score` value, :math:`trace(APB^T P^T)`, can be calculated by forming
+    the permutation matrix:
+    >>> n = len(A)
+    >>> P = np.zeros((n, n), dtype=int)
+    >>> P[np.arange(n), res.col_ind] = 1
+    >>> score = np.trace(A @ P @ B.T @ P.T)
+    >>> print(score)
+    200.0
+
+    Alternatively, the score can be calculated by applying the appropriate
+    permutation of ``B``:
+    >>> score = np.trace(A @ B[res.col_ind][:, res.col_ind].T)
+    >>> print(score)
+    200.0
+    """
+    A = np.asarray(A)
+    B = np.asarray(B)
+
+    if A.ndim != 2:
```

Comment:
This doesn't check for squareness, right? It could be rectangular.

In any case, I'm going to merge this with gh-11862, so we'll take care of input validation there.

Reply:
Whoops. You are right.
I have added a check for squareness all the same so I can mark this as resolved.